[FEAT] 로그아웃 기능 구현

dye0p opened the pull request:
## Issue Number
closed #471

## As-Is
<!-- 문제 상황 정의 -->
로그아웃 기능의 부재로 사용자는 로그아웃을 할 수 없었습니다.

## To-Be
<!-- 변경 사항 -->
인증된 사용자는 로그아웃을 할 수 있습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- 신기능
  - 인증된 사용자를 위한 로그아웃 엔드포인트 추가: 로그아웃 시 액세스/리프레시 토큰 쿠키가 즉시 삭제되며 응답은 콘텐츠 없이 완료됩니다.
- 버그 수정
  - 인증 정보가 누락된 경우 조기 검증으로 명확한 오류를 반환해 예기치 않은 동작을 방지합니다.
- 테스트
  - 로그아웃 시 쿠키 삭제 및 리프레시 토큰 제거를 검증하는 통합/단위 테스트 추가로 안정성 강화.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dye0p committed:
feat: RefreshToken 엔티티 getter 클래스 레벨에 선언

dye0p committed:
feat: 비어있는 accessToken과 refreshToken을 쿠키에 write 하는 기능 구현

dye0p committed:
feat: 만료시간이 0인 비어있는 쿠키를 만드는 기능 구현

dye0p committed:
feat: 로그아웃 서비스 로직 구현

- 로그아웃시 리프레시 토큰을 db에서 제거한다.

dye0p committed:
feat: 로그아웃 API 구현

- 로그아웃시 현재 로그인된 사용자의 인가정보를 기반으로 로그아웃을 진행한다.
- 로그아웃시 쿠키에 포함될 토큰은 비어있는 토큰으로 덮어씌운다.

dye0p committed:
feat: resolver memberId null 체크 예외 구현

dye0p committed:
feat: memberId로 refreshToken을 찾는 쿼리메서드 추가

dye0p committed:
test: 로그아웃 service 로직 테스트 추가

dye0p committed:
test: 로그아웃 통합 테스트 추가

dye0p commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationArgumentResolver.java:
```diff
@@ -20,10 +20,17 @@ public boolean supportsParameter(MethodParameter parameter) {
         HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();
         String requestAttributeName = "memberId";
         Object memberId = request.getAttribute(requestAttributeName);
+        if (memberId == null) {
+            throw new NullPointerException("memberId가 null 입니다."); //403 or 500 논의
+            /*
+            403: 개발자의 실수로 AuthRequired를 안붙인 것은 맞지만, 클라이언트의 입장에서는 인가가 필요한 상황인데 인가가 되지 않은 상황이므로 401
+            500: 서버쪽에서 예상하지 못한 버그 같은 것들, 500으로 할거면 어떤 예외를 할지 고민
+             */
+        }
```
일전에 띠용과 얘기를 해봤지만 계속 헷갈려서 논의 필요할 것 같아요!

jye326 replied:
- 400 번대 에러는 클라이언트 오류, 500 번대 에러는 서버 오류입니다.
- 403 Forbidden 오류는 인증은 통과되었으나 인가 받지 않은, 정책적으로 금지된 요청을 보낼 때 발생합니다.
- 해당 상황에서 클라이언트는 과실이 전혀 없습니다. 100% 서버 에러(개발자 휴먼 에러)로 500이 맞다고 생각합니다.

kimdevv replied:
이 오류는 개발자의 실수로 `@AuthRequired`를 빠뜨린 상황에만 발생합니다. 401이나 403 응답을 받더라도 사용자가 취할 수 있는 조치가 없습니다. 100% 서버 에러이므로 저도 500이라고 생각합니다.

ppparkta replied:
서버 에러인 쪽에 동의합니다

dye0p replied:
NullPointException을 던지도록 하였고, 해당 예외는 500 상태코드를 반환합니다!

[refactor: Resolver 주석제거 및 validation 메서드 분리](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/0ae5732a77ac6d97a3abeed9c218e0d538181186)

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/CookieProvider.java:
```diff
@@ -12,4 +12,14 @@ public static ResponseCookie createCookie(final String name, final String value)
                 .build();
     }
+
+    public static ResponseCookie clearCookie(final String name) {
+        return ResponseCookie.from(name, "")
+                .httpOnly(true)
+                .secure(true)
+                .path("/")
+                .sameSite("None")
+                .maxAge(0)
+                .build();
```
p3 : createCookie 와 많은 부분이 중복되는 것 같은데 return create(name, "") 로 구현하지 않고, 별도로 작성하신 이유가 궁금합니다!

dye0p replied:
별도의 메서드로 분리했던 이유는 쿠키의 만료시간 설정 때문이었습니다. 
- `.maxAge(0)`
- 토큰을 가지는 쿠키는 만료시간이 토큰 자체에 있기 때문에 별도의 만료시간이 필요없었지만, 빈 값을 가지는 쿠키는 만료시간을 0으로 설정해야 했습니다. 

하지만, 말씀해주신 것 처럼 중복된 부분들이 많다는 부분은 동의합니다! 그래서 빌더 생성자를 사용하도록 리팩터링 하였습니다!

[refactor: ResponseCookie 빌더 생성자 추가](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/4557c99a55cd85b89c8aff71debf45051a36bdd8)

jye326 replied:
좋은데요~

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/CookieWriter.java:
```diff
@@ -17,6 +17,14 @@ public static void write(HttpServletResponse response, AuthTokenResponse tokens)
         response.addHeader(HttpHeaders.SET_COOKIE, refreshToken.toString());
     }
+
+    public static void write(HttpServletResponse response) {
+        ResponseCookie accessToken = CookieProvider.clearCookie("accessToken");
+        ResponseCookie refreshToken = CookieProvider.clearCookie("refreshToken");
+
+        response.addHeader(HttpHeaders.SET_COOKIE, accessToken.toString());
+        response.addHeader(HttpHeaders.SET_COOKIE, refreshToken.toString());
+    }
```
p2 : 
`write(HttpServletResponse response, AuthTokenResponse tokens)` : 해당 토큰 값을 쿠키에 쓰는 메서드
`write(HttpServletResponse response)` : 쿠키를 초기화 하는 메서드로 이해했습니다.

- 쿠키 내부 토큰 초기화 메서드는 `write` 보다 `clear`인 편이 혼동을 줄일 수 있을 것 같습니다. 쿠키에 쓸 값을 입력 받지도 않고, 그냥 초기화만 하는 메서드니까요. 
- 실제로 쿠키에 유효한 토큰 값을 쓰는 기존 write 메서드와 역할이 명확히 구분되어 있으므로 굳이 오버로딩할 필요는 없어보여요!

dye0p replied:
> 쿠키 내부 토큰 초기화 메서드는 write 보다 clear인 편이 혼동을 줄일 수 있을 것 같습니다. 쿠키에 쓸 값을 입력 받지도 않고, 그냥 초기화만 하는 메서드니까요.

동의합니다! 서로 다른 역할을 하는 메서드이므로 네이밍으로 구별해주는게 좋을 것 같아요. 반영하겠습니다.

[refactor: 빈 쿠키를 쓰는 메서드 네이밍 명확하게 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/e4fa0c0382b32364ebe016c5216e2226b3e18af3)

jye326 reviewed (commented):
이번에도 테스트를 꼼꼼히 작성해 주셨군요! 👍
수고하셨습니다~ 코멘트 확인 한번 부탁드려요!

kimdevv reviewed (approved):
고생하셨습니다~ 이미 좋은 커멘트들이 남겨져 있어서 저는 이만 승인합니다

ppparkta commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationArgumentResolver.java:
```diff
@@ -20,10 +20,17 @@ public boolean supportsParameter(MethodParameter parameter) {
+            /*
+            403: 개발자의 실수로 AuthRequired를 안붙인 것은 맞지만, 클라이언트의 입장에서는 인가가 필요한 상황인데 인가가 되지 않은 상황이므로 401
+            500: 서버쪽에서 예상하지 못한 버그 같은 것들, 500으로 할거면 어떤 예외를 할지 고민
+             */
```
p3: 논의 필요한 내용 주석으로 다니까 놓치지 않고 보게돼서 좋네요. 머지되기 전에 잊지 말고 지워만 주세요!

dye0p replied:
논의 결과대로 수정하였습니다!

[refactor: Resolver 주석제거 및 validation 메서드 분리](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/0ae5732a77ac6d97a3abeed9c218e0d538181186)

ppparkta reviewed (commented):
고생하셨습니다. 논의 중인 내용 답변 달리면 승인하겠습니다!

dye0p commented:
@coderabbitai 리뷰해

dye0p committed:
refactor: Resolver 주석제거 및 validation 메서드 분리

dye0p committed:
refactor: 빈 쿠키를 쓰는 메서드 네이밍 명확하게 변경

- 기존의 write는 토큰을 담은 쿠키를 쓰는 write 메서드와 혼동을 줄 수 있다고 생각해서 명확한 네이밍으로 변경

dye0p committed:
refactor: ResponseCookie 빌더 생성자 추가

dye0p committed:
refactor: 로그아웃 API HTTP 메서드 변경(get -> post)

- 로그아웃시 기존의 토큰을 삭제하고 새로운 쿠키를 쓰는등의 작업이 있으므로 조회 시 많이 사용되는 get보단, 상태 변화 API에 적합한 post로 변경

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -43,6 +47,14 @@ public ResponseEntity<Void> login(@RequestBody @Valid SignInRequest request, Htt
+        authService.logout(loginInfo.memberId());
+        CookieWriter.write(httpResponse);
+        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+    }
```
post로 변경완료! 

[refactor: 로그아웃 API HTTP 메서드 변경(get -> post)](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/1d38483655ebcaf97671105087890d9ae1345fb7)

dye0p committed:
fix: 리프레시 토큰 삭제시 deleteAll() 사용하도록 수정

- 기존의 findBy는 단건의 리프레시 토큰을 삭제 했지만, 여러 memberId에 대한 refreshToken이 있을 수 있으므로 모든 memberId에 대한 토큰을 제거할 수 있도록 수정

dye0p committed:
test: 로그아웃 E2E 테스트 결과 쿠키값 세부내역 검증 추가

dye0p committed:
test: 존재하지 않는 토큰에 대한 로그아웃 멱등성 테스트 추가

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/RefreshTokenRepository.java:
```diff
@@ -9,4 +9,6 @@
 
     Optional<RefreshToken> findByTokenValue(String token);
+
+    Optional<RefreshToken> findByMemberId(Long memberId);
```
멱등성 보장과 멱등성 보장 테스트 추가 완료!

[fix: 리프레시 토큰 삭제시 deleteAll() 사용하도록 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/b50a6f22f1b7c4f36831da01e8ee0314715c1591)

[test: 존재하지 않는 토큰에 대한 로그아웃 멱등성 테스트 추가](https://github.com/woowacourse-teams/2025-Fit-toring/pull/473/commits/cb298fc92ba91394e3ad3c516f3d1bc0cf121008)
